### PR TITLE
Continued subj sequential selection

### PR DESCRIPTION
**Pull request by yuenmichelle1**
Drafting PR. Currently still testing this. Not too much different than previous PR created by Amy and Zach (maybe an added test here and there).

### CODE REVIEW

**Comment by yuenmichelle1** [comment]
@camallen  I mentioned this on message but absolutely no rush on this. This is still a work in progress  but would appreciate some  feedback/insights if i am missing certain aspects whenever you have a moment.

**Review by camallen — COMMENTED** [review]
This looks really good and pretty close i'd say. I'm mostly wondering how the `set_member_subject` db data loading and null `priority` column values work with the proposed `keysort` approach. 

It would be nice to run this against some fake data and check the random selector still distributes the selected subject ids randomly (i.e. what's the resulting distribution of selected values after say 100 selection runs).... we don't want a regression in random as it's our main feature in this repo. I assume this won't be a case but i think it'll be good to check our assumptions as the test code doesn't really test this (it's really hard to test random output), we can do this latter though.

Nice work @yuenmichelle1 👏🏼

File: web/models/subject.ex [hunk 1]
```diff
@@ -1,20 +0,0 @@
-defmodule Designator.Subject do
```
[the remark this answers is not in the document]
Comment (yuenmichelle1) [reply web/models/subject.ex:1]: Correct. It doesnt look like it's been used. Looking at previous conversations between Zach and Marten, it seems like originally it was thought Subjects were going to be used but as code gets built out more , discovery was that more was done with Workflow (and Set Member Subject side) vs Subjects. I looked around and it doesnt look like  Subject module is being used anywhere else.

File: lib/designator/subject_set_iterators/randomly.ex [hunk 2]
```diff
@@ -1,8 +1,8 @@
-defmodule Designator.RandomStream do
+defmodule Designator.SubjectSetIterators.Randomly do
```
[the remark this answers is not in the document]
Comment (yuenmichelle1) [reply lib/designator/subject_set_iterators/randomly.ex]: @camallen  Sounds good. I'm good with making this change. I'm wondering if there could be potential future confusion since there is `Designator.Random` module in place already, but it may not be too big of a deal since one is under `Designator.SubjectSetIterators`.

File: lib/designator/subject_set_iterators/sequentially.ex [hunk 3]
```diff
@@ -0,0 +1,14 @@
+defmodule Designator.SubjectSetIterators.Sequentially do
+  @spec apply_to(Enumerable.t) :: Enumerable.t
+  def apply_to(enum) do
+    Stream.with_index(enum) |> Stream.map(fn {elem,index} -> {index, elem} end)
+  end
+
+  defp size(enum = %Array{}) do
```
[the remark this answers is not in the document]
Comment (yuenmichelle1) [reply lib/designator/subject_set_iterators/sequentially.ex]: For `subject_set_iterators/sequential.ex`  no, but yes for `subject_set_iterators/random.ex` (`random.ex` using its own private `size` method). Initially, my thought was to keep the `SubjectSetIterators`  similar, but you are right, it's not being used, so I'll remove.

File: web/models/workflow.ex [hunk 4]
```diff
@@ -27,8 +28,10 @@ defmodule Designator.Workflow do
     query = from sms in "set_member_subjects",
       left_join: swc in "subject_workflow_counts", on: (sms.subject_id == swc.subject_id and swc.workflow_id == ^workflow_id),
       where: sms.subject_set_id == ^subject_set_id and is_nil(swc.retired_at),
-      select: sms.subject_id
+      select: {sms.subject_id, sms.priority}
```
[the remark this answers is not in the document]
Comment (yuenmichelle1) [reply web/models/workflow.ex:31]: You're right, I should add test cases for null case. From what I can tell (correct me if i'm wrong), if `priority` is null `keysort` will spit out the results of the query as it was given from the db (assuming all the `set_member_subjects` from the query have `priority` null. If db returns  `set_member_subjects` where some have `priority` and some are null,  then it will first sort those that do have priority then append null priority `set_member_subjects` in the order given by db.

And you are right that this might affect randomness as it was working before, which isn't my intention and missed here. The more it's pointed out, the more I'm thinking I should move the sort by priority to `Sequential` module. 

Thank you for catching! @camallen

[the remark this answers is not in the document]
Comment (camallen) [reply web/models/workflow.ex:31]: > if priority is null keysort will spit out the results of the query as it was given from the db 

I assume so from the elixir docs (the sort is stable) https://hexdocs.pm/elixir/1.1.1/List.html#keysort/2

> (assuming all the set_member_subjects from the query have priority null.

In theory we should have priority for either all the subjects or none of the subjects but that's an ideal world :) It's currently not so easy to get the priority values consistently uploaded by researchers so we can end up with mixes of values being set on or missing.

> then it will first sort those that do have priority then append null priority set_member_subjects in the order given by db.

Yep - that sounds right. In theory for most of our random workflows this won't make a difference if the sort is stable and preserves the ordering from the DB (same behaviour as before) which is good to know but we should test it.

> The more it's pointed out, the more I'm thinking I should move the sort by priority to Sequential module.

TBH - if we can test/verify the ordering preservation on data loading for both cases (workflows for prioirty are set or missing) then we can most likely use this one implementation which seems simpler to me. 

I quite like this implementation as long as it doesn't change the overall system behaviour as it is simpler, load data and apply the needed transform for the downstream selection strategy (random vs sequential).

[the remark this answers is not in the document]
Comment (yuenmichelle1) [reply web/models/workflow.ex:31]: > TBH - if we can test/verify the ordering preservation on data loading for both cases (workflows for priority are set or missing) then we can most likely use this one implementation which seems simpler to me.

Just to be clear, this would have to be a test for a staging environment, correct? I've added tests for null state cases on workflow model file test, but I think this is ready for full review. Let me know if I'm missing anything. I definitely will need to test this super rigorously to make sure these changes don't break what's currently working.

[the remark this answers is not in the document]
Comment (camallen) [reply web/models/workflow.ex:31]: there are a few ways we could do this - i'll reach out directly to discuss.

**Comment by yuenmichelle1** [comment]
[after.txt](https://github.com/zooniverse/designator/files/6888445/after.txt)
[before.txt](https://github.com/zooniverse/designator/files/6888446/before.txt)
[random_output.xlsx](https://github.com/zooniverse/designator/files/6888447/random_output.xlsx)
Testing random selection to ensure similar behavior as master. (Output in txt created using attached bash script) 
[designator_random_test.txt](https://github.com/zooniverse/designator/files/6888454/designator_random_test.txt)

**Comment by camallen** [comment]
> Behavior looks fairly similar used workflow 18975 as test.

I only see the raw data - Did you create a histogram of each subject and how many times it was selected randomly, for both the before and after? Am i missing a results file?

**Comment by yuenmichelle1** [comment]
> > Behavior looks fairly similar used workflow 18975 as test.
> 
> I only see the raw data - Did you create a histogram of each subject and how many times it was selected randomly, for both the before and after? Am i missing a results file?

Oh gosh. My bad. I uploaded the wrong xlsx file. 
[random_resultswf18795.xlsx](https://github.com/zooniverse/designator/files/6901050/random_resultswf18795.xlsx)
Let me know if you see any issues.

**Comment by yuenmichelle1** [comment]
> > > Behavior looks fairly similar used workflow 18975 as test.
> > 
> > 
> > I only see the raw data - Did you create a histogram of each subject and how many times it was selected randomly, for both the before and after? Am i missing a results file?
> 
> Oh gosh. My bad. I uploaded the wrong xlsx file.
> [random_resultswf18795.xlsx](https://github.com/zooniverse/designator/files/6901050/random_resultswf18795.xlsx)
> Let me know if you see any issues.

@camallen  Thanks for looking at these. I appreciate your feedback! :) I'm in no rush and want to make sure everything is working right. 

[updated_random_resultsWF18795.xlsx](https://github.com/zooniverse/designator/files/6902186/updated_random_resultsWF18795.xlsx)
Updated histogram data to display per id frequency for before and after. Previous histogram automatically grouped ids by ranges for easier table columns, instead we want a per id view. 

Looking at both histograms it doesn't look like frequency is skewing to any id. You mentioned on slack before, frequency for each id revolves around 40s, 50s, 60s range for before and  also the after which is pretty good. Maybe one thing that i noticed is that the id of the last subject shows up less in the after version as opposed to the before (frequency of 36 vs 62), but I don't think that's too significant (if we are rounding ~40 vs ~60). I should forewarn you that statistical analysis isn't my forte (my surprising flaw as a math major), so if you see anything that stands out, let me know.

**Comment by camallen** [comment]
> Looking at both histograms it doesn't look like frequency is skewing to any id. You mentioned on slack before, frequency for each id revolves around 40s, 50s, 60s range for before and also the after which is pretty good. Maybe one thing that i noticed is that the id of the last subject shows up less in the after version as opposed to the before (frequency of 36 vs 62), but I don't think that's too significant (if we are rounding ~40 vs ~60). I should forewarn you that statistical analysis isn't my forte (my surprising flaw as a math major), so if you see anything that stands out, let me know.

The results look good - it seems the `after` data set was 5 values as the total counts of selected ids was 4995 `after` vs 5000 `before`.  This would most likely be due to a missing a pre-warmed cache on the localhost test machine as the data loads on the first request and you get back and empty `[]` instead of 5 selected ids. Note 5 is the default selection size https://github.com/zooniverse/designator/blob/3ec13a83a13ff0a54fd60e1911bade46c863099c/web/controllers/workflow_controller.ex#L10

### From the excel file
| | Before | After |  
|-|-|-|
Average | 50 | 49.95
Median | 51 | 55
Mode | 50 | 50

Noting that this is a random selection strategy so we should see some slight variation in these distributions and numbers at the smaller selection repeats (100 by default). 

To me both the `before` and `after` distribution looks pretty much the same and they have very similar average, median and mode selection values. I dare say if we ran the selections more than 100 times (perhaps 500 / 1000) we'd see the small discrepancy shrink away and the distribution of selected ids would better match.

OOI - did you test the sequential feature we're implementing here on a known prioritized workflow?

File: test/models/workflow_test.exs [hunk 5]
```diff
@@ -40,15 +40,25 @@ defmodule Designator.WorkflowTest do
       (1, NOW(), NOW()),
       (2, NOW(), NOW())")
       Ecto.Adapters.SQL.query!(Designator.Repo, "INSERT INTO subject_sets_workflows (workflow_id, subject_set_id) VALUES (1, 1)")
-      Ecto.Adapters.SQL.query!(Designator.Repo, "INSERT INTO set_member_subjects (subject_set_id, subject_id, random, created_at, updated_at) VALUES
-      (1, 1, 0.5, NOW(), NOW()),
-      (1, 2, 0.5, NOW(), NOW()),
-      (1, 3, 0.5, NOW(), NOW())")
+      Ecto.Adapters.SQL.query!(Designator.Repo, "INSERT INTO set_member_subjects (subject_set_id, subject_id, priority, random, created_at, updated_at) VALUES
+      (1, 1, 3, 0.5, NOW(), NOW()),
```
[the remark this answers is not in the document]
Comment (camallen) [reply test/models/workflow_test.exs:44]: I would remove the priority values in these some of these non-sequential feature tests to make sure the system handles the expected null values being present in the database.

File: web/models/workflow.ex [hunk 6]
```diff
@@ -27,8 +28,10 @@ defmodule Designator.Workflow do
...
     Designator.Repo.all(query)
+    |> List.keysort(1)
```
[the remark this answers is not in the document]
Comment (camallen) [reply web/models/workflow.ex:33]: the sort is a stable sort so will respect the original data ordering as it comes out of the database when the priority values are missing, see https://github.com/zooniverse/designator/pull/117/files#r674803948

**Review by camallen — APPROVED** [review]
I think this is ready to get to the staging system for testing. 

I think you should remove the priority values in the tests that don't require them as that will also test the stability of the sorting algorithm with null values. I've added a comment outlining this spot (basically additions where they old tests should have been fine to leave as is).

Nice work :) 👏🏼

**Comment by yuenmichelle1** [comment]
> OOI - did you test the sequential feature we're implementing here on a known prioritized workflow?

Sorry for late response. I tested with a couple workflows 18861, 18814, 18805 (related to Davy Notebooks) where `prioritized` is `true` and subjects are coming up in sequential order as we want, which is nice.

**Comment by yuenmichelle1** [comment]
Thanks Cam!
